rllib/utils/replay_buffers/episode_replay_buffer.py

```diff
@@ -0,0 +1,508 @@
+from collections import deque
+from typing import Any, Dict, Optional
+import uuid
+
+import numpy as np
+
+from ray.rllib.policy.sample_batch import MultiAgentBatch, SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.replay_buffers.replay_buffer import ReplayBufferInterface
+from ray.rllib.utils.typing import SampleBatchType
+
+
+class EpisodeReplayBuffer(ReplayBufferInterface):
+    """Buffer that stores (completed or truncated) episodes by their ID.
+
+    Each "row" (a slot in a deque) in the buffer is occupied by one episode. If an
+    incomplete episode is added to the buffer and then another chunk of that episode is
+    added at a later time, the buffer will automatically concatenate the new fragment to
+    the original episode. This way, episodes can be completed via subsequent `add`
+    calls.
+
+    Sampling returns batches of size B (number of "rows"), where each row is a
+    trajectory of length T. Each trajectory contains consecutive timesteps from an
+    episode, but might not start at the beginning of that episode. Should an episode end
+    within such a trajectory, a random next episode (starting from its t0) will be
+    concatenated to that "row". Example: `sample(B=2, T=4)` ->
+
+       0 .. 1 .. 2 .. 3  <- T-axis
+    0 e5   e6   e7   e8
+    1 f2   f3   h0   h2
+    ^ B-axis
+
+    .. where e, f, and h are different (randomly picked) episodes, the 0-index (e.g. h0)
+    indicates the start of an episode, and `f3` is an episode end (gym environment
+    returned terminated=True or truncated=True).
+
+    0-indexed returned timesteps contain the reset observation, a dummy 0.0 reward, as
+    well as the first action taken in the episode (action picked after observing
+    obs(0)).
+    The last index in an episode (e.g. f3 in the example above) contains the final
+    observation of the episode, the final reward received, a dummy 0-action, as well
+    as either terminated=True or truncated=True.
+    """
+    def __init__(
+        self,
+        capacity: int = 10000,
+        *,
+        batch_size_B: int = 16,
+        batch_length_T: int = 64,
+    ):
+        """Initializes an EpisodeReplayBuffer instance.
+
+        Args:
+            capacity: The total number of timesteps to be storable in this buffer.
+                Will start ejecting old episodes once this limit is reached.
+            batch_size_B: The number of rows in a SampleBatch returned from `sample()`.
+            batch_length_T: The length of each row in a SampleBatch returned from
+                `sample()`.
+        """
+        self.capacity = capacity
+        self.batch_size_B = batch_size_B
+        self.batch_length_T = batch_length_T
+
+        # The actual episode buffer.
+        self.episodes = deque()
+        # Maps (unique) episode IDs to the index under which to find this episode
+        # within our `self.episodes` deque. Note that after ejection started, the
```

Comment:
I think this is commonly called "_eviction_".

Reply:
fixed